### DISCUSSION TITLE
How to make constant values show up in search results?

### DISCUSSION THREAD

**Discussion by nktnet1** [fuma-nama/fumadocs#806] [Q&A] [answered]
E.g. in `constants.ts`:

```ts
export const EXAMPLE = 'helloworld';
```

Then in `example-content.mdx`:
```mdx
import { EXAMPLE } from './constants';

This variable will not be indexed for search: {EXAMPLE}
```

Is there a seamless way to make this constant value appear in searches without needing to tweak the starter search code?

Thanks!

---

**Update:** should also add that searching for **EXAMPLE** matches (false positive), and searching for "helloworld" does not match (false negative).

**Comment by fuma-nama** [comment]
No. The easiest way is to create a hidden element with the actual text value.

**Reply by nktnet1** [reply]
I see... and I take it that there's no way to make searching for the internal variable name to not show up either?

My goal is to re-use the same value across different pages, so I think adding hidden elements with the correct text will defeat that purpose.

**Reply by fuma-nama** [reply]
Using a remark plugin would be great for this purpose, it works perfectly because search indexes are built during build time on the bundler. But rendering/importing the constant requires Next.js runtime, the bundler doesn't know the actual value of constants at build time.

**Reply by nktnet1** [reply]
Thanks for the follow up @fuma-nama

> Using a remark plugin would be great for this purpose

Is "purpose" in this case the ability to re-use the same value across different pages and still appear in searches? And are you saying that this is possible if I install a remark-plugin?

If so, would it be okay for you to point me in the right directions regarding the plugin?

Otherwise, if it isn't possible, no worries! Thanks!

**Reply by fuma-nama** [reply]
Really wanted to make a tutorial about writing remark plugins but I don't have time. 

Basically in MAST, usages of constant value resolves to:

```mdx
{test}
```

```json
{
  "type": "root",
  "children": [
    {
      "type": "mdxFlowExpression",
      "value": "test",
      "data": {
        "estree": {
          "type": "Program",
          "body": [
            {
              "type": "ExpressionStatement",
              "expression": {
                "type": "Identifier",
                "name": "test"
              }
            }
          ],
          "sourceType": "module",
          "comments": []
        }
      }
    }
  ]
}
```

You can make a remark plugin, use `unist-util-visit` to walk through the tree, and replace `mdxFlowExpression` of a certain value with your desired node. Like a text node:

```json
{
  "type": "text",
  "value": "some value here"
},
```

**Reply by nktnet1** [reply, marked as answer]
I figured out how to do it based on your suggestion. Thank you!

First, install the following:

```sh
pnpm i -D unist-util-visit @types/mdast
```

In `source.config.ts`, create the following plugin:

```ts
import type { Root } from "mdast";
import { visit } from "unist-util-visit";

function remarkReplaceConstants(options: {
  constants: Record<string, string>;
}) {
  return (tree: Root) => {
    visit(tree, (node, index, parent) => {
      if (
        node.type !== "mdxFlowExpression"
        && node.type !== "mdxTextExpression"
      ) {
        return;
      }

      const replacementValue = options.constants[node.value];
      if (
        parent
        && replacementValue !== undefined
        && typeof index === "number"
      ) {
        parent.children[index] = {
          type: "text",
          value: replacementValue,
        };
      }
    });
  };
}
```

then use it as follows:

```ts
import { defineConfig } from "fumadocs-mdx/config";

export default defineConfig({
  mdxOptions: {
    remarkPlugins: [
      [
        remarkReplaceConstants,
        {
          constants: {
            APP_NAME: "Webview Kiosk",
            // Other constants
          },
        },
      ],
    ],
  },
});
```

In MDX, we can now use `{APP_NAME}` without importing, and it will be automatically replaced.

Complete Example:

- https://github.com/nktnet1/webview-kiosk/pull/250
